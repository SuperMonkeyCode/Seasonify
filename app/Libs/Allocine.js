/* @flow */
/* eslint-disable no-use-before-define */
import axios from 'axios';
import jshashes from 'jshashes';

import type { ApiResponse, ApiPromise } from 'Types';
import DebugConfig from 'Config/DebugConfig';

// DOC : https://wiki.gromez.fr/dev/api/allocine_v3
export class Allocine {
  config = {
    apiHostName: 'http://api.allocine.fr',
    apiBasePath: '/rest/v3/',
    apiPartner: '100043982026',
    apiSecretKey: '29d185d98c984a359e6e6f26a0474269',
    imgBaseUrl: 'http://images.allocine.fr',
  };

  presets = {
    global: {
      format: 'json',
      partner: this.config.apiPartner,
    },
    search: { filter: 'tvseries', count: 5 },
    tvseries: { profile: 'large' },
    season: { profile: 'medium' },
  };

  sha1 = new jshashes.SHA1();

  static removeDynamicData(field: string, data: Array<Object>): Array<Object> {
    return data.map(tvshow => {
      const { [field]: deleted, ...newTvShowContent } = tvshow;

      return { ...newTvShowContent };
    });
  }

  todayDate = (): string => {
    const date = new Date();
    return (
      date.getFullYear() +
      ('0' + (date.getMonth() + 1)).slice(-2) + // eslint-disable-line prefer-template
      ('0' + date.getDate()).slice(-2) // eslint-disable-line prefer-template
    );
  };

  buildPath = (route: string, specificParams: Object) => {
    const Url: string = this.config.apiBasePath + route;

    const params: Object = Object.assign(
      {},
      this.presets.global,
      this.presets[route],
      specificParams,
    );

    const paramsString: string = Object.keys(params)
      .map(key => [key, params[key]].map(encodeURIComponent).join('='))
      .join('&');

    const sed: string = this.todayDate();

    const sig: string = `${this.config.apiSecretKey}${paramsString}&sed=${sed}`;
    const encodedSig: string = encodeURIComponent(this.sha1.b64(sig));

    return `${Url}?${paramsString}&sed=${sed}&sig=${encodedSig}`;
  };

  get = (method: string, options: Object): Promise<*> => {
    const path: string = this.buildPath(method, options);
    /* istanbul ignore next */
    if (DebugConfig.logApiCallToConsole) {
      console.log('Allocine call: ', this.config.apiHostName + path); // eslint-disable-line no-console
    }
    return axios.get(this.config.apiHostName + path);
    // .then(result => ({ error: null, data: result.data }))
    // .catch(error => ({ error: error.message, data: null }));
  };
}

const api = new Allocine();

/* ========== API CALLS ========== */

export default {
  class: api,
  searchTvshows: (query: string): ApiResponse | ApiPromise => {
    if (query.trim() === '') {
      return { error: null, data: [] };
    }
    return api
      .get('search', { q: query })
      .then(result => ({
        error: null,
        data: result.data.feed && result.data.feed.tvseries
          ? Allocine.removeDynamicData('statistics', result.data.feed.tvseries)
          : [],
      }))
      .catch(error => ({ error: error.message, data: null }));
  },

  getSeasons: (code: number): ApiPromise =>
    api
      .get('tvseries', { code })
      .then(result => {
        if (!result.data.tvseries) {
          return { error: "API didn't found the TV show", data: null };
        }
        const data: Array<Object> = Allocine.removeDynamicData(
          'statistics',
          result.data.tvseries.season,
        );
        return {
          error: null,
          data,
        };
      })
      .catch(error => ({ error: error.message, data: null })),
};
