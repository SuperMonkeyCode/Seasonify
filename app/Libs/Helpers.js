/* @flow */
import type { SortingValue } from 'Types';

export function trimmed(subject: any) {
  if (typeof subject === 'string') {
    return subject.trim();
  } else if (typeof subject === 'object' && 'map' in subject) {
    return subject.map(item => (typeof item === 'string' ? item.trim() : item));
  } else if (typeof subject === 'object') {
    return Object.keys(subject).reduce((newObj, key) => {
      newObj[key] =
        typeof subject[key] === 'string' // eslint-disable-line no-param-reassign
          ? subject[key].trim()
          : subject[key];
      return newObj;
    }, {});
  }
  return subject;
}

export function collecToArray(collection: Object): Array<Object> {
  return Object.keys(collection).map(key => ({
    id: key,
    ...collection[key],
  }));
}

export function getContrastingTextColor(hex: string) {
  /*
    From this W3C document: http://www.webmasterworld.com/r.cgi?f=88&d=9769&url=http://www.w3.org/TR/AERT#color-contrast
    Color brightness is determined by the following formula:
    ((Red value X 299) + (Green value X 587) + (Blue value X 114)) / 1000
    */

  function hexToR(h) {
    return parseInt(cutHex(h).substring(0, 2), 16);
  }
  function hexToG(h) {
    return parseInt(cutHex(h).substring(2, 4), 16);
  }
  function hexToB(h) {
    return parseInt(cutHex(h).substring(4, 6), 16);
  }
  function cutHex(h) {
    return h.charAt(0) === '#' ? h.substring(1, 7) : h;
  }

  const threshold = 130; /* about half of 256. Lower threshold equals more dark text on dark background  */

  const hRed = hexToR(hex);
  const hGreen = hexToG(hex);
  const hBlue = hexToB(hex);

  const cBrightness = (hRed * 299 + hGreen * 587 + hBlue * 114) / 1000;
  if (cBrightness > threshold) {
    return '#000000';
  }
  return '#ffffff';
}

export function sortAlpha(sorting: SortingValue, a: string, b: string): number {
  if (sorting === 'DESC') return b.localeCompare(a);
  return a.localeCompare(b);
}
