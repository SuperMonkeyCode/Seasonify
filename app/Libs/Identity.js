import uuid from 'uuid';

export default class Identity {
  static lastId = 0;
  static forcedIds = [];

  static newid() {
    if (this.forcedIds.length > 0) {
      this.lastId = this.forcedIds.shift();
    } else {
      this.lastId = uuid();
    }
    return this.lastId;
  }

  static id() {
    return this.lastId;
  }

  static forceId(ids) {
    if (!Array.isArray(ids)) {
      this.forcedIds = [ids];
      return ids;
    }
    this.forcedIds = ids;
    return ids[0];
  }
}
