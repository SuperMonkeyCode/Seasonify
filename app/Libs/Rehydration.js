import { AsyncStorage } from 'react-native';
import { persistStore } from 'redux-persist';
import PersistConfig from 'Config/PersistConfig';
import ui from 'State/uiState';

const updateReducers = (store: Object) => {
  const reducerVersion = PersistConfig.reducerVersion;
  const config = PersistConfig.storeConfig;
  const start = () => store.dispatch(ui.actions.startup());

  // Check to ensure latest reducer version
  AsyncStorage.getItem('reducerVersion')
    .then(localVersion => {
      if (localVersion !== reducerVersion) {
        console.tron.display({
          name: 'PURGE',
          value: {
            'Old Version:': localVersion,
            'New Version:': reducerVersion,
          },
          preview: 'Reducer Version Change Detected',
          important: true,
        });
        // Purge store
        persistStore(store, config, start).purge();
        AsyncStorage.setItem('reducerVersion', reducerVersion);
      } else {
        persistStore(store, config, start);
      }
    })
    .catch(() => {
      persistStore(store, config, start);
      AsyncStorage.setItem('reducerVersion', reducerVersion);
    });
};

export default { updateReducers };
