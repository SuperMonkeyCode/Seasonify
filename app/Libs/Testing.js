import deepFreeze from 'deep-freeze';

/*
testReducer(
  reducer,
  [
    actions.messageToast('success', 'Message'),
    actions.messageToast('error', 'Other message'),
  ],
  {
    message: {
      type: 'error',
      text: 'Other message',
    },
  },
 */
export function testReducer(reducer, action, expectedState) {
  const initialState = reducer(undefined, {});
  deepFreeze(initialState);
  deepFreeze(expectedState);
  let stateType = 'non-object';
  if (Array.isArray(initialState)) stateType = 'array';
  else if (typeof state === 'object') stateType = 'object';

  // if single action, still put in an array
  const actions = Array.isArray(action) ? action : [action];

  // run all the actions/reducers and combine de states. We've to account for that fact that the initialState can be a literal, an object or an array
  const finalState = Object.keys(actions).reduce((state, key) => {
    const act = actions[key];
    switch (stateType) {
      case 'array':
        return [...state, ...reducer(state, act)];
      case 'object':
        return { ...state, ...reducer(state, act) };
      default:
        return reducer(state, act);
    }
  }, initialState);

  expect(finalState).toEqual({ ...initialState, ...expectedState });
}
