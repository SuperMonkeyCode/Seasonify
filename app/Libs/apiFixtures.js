/* @flow */

export default {
  searchTvshows: (query: string): ApiResponse => {
    if (query.trim() === '') {
      return { error: null, data: [] };
    }
    return {
      error: null,
      data: require('Fixtures/apiSearchHappy.json'),
    };
  },

  getSeasons: (): ApiResponse => ({
    error: null,
    data: require('Fixtures/apiSeasonsDeadwood.json'),
  }),
};
