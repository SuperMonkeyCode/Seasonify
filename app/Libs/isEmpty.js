/* @flow */
export default function isEmpty(obj: any): boolean {
  if (obj == null) return true;
  if (Array.isArray(obj) || typeof obj === 'string') return obj.length === 0;
  // istanbul ignore next
  // eslint-disable-next-line no-prototype-builtins, no-restricted-syntax
  for (const key in obj) if (obj.hasOwnProperty(key)) return false;
  return true;
}
