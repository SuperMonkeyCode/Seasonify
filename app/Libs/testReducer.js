import deepFreeze from 'deep-freeze';

export default function testReducer(reducer, init, action, expectedState) {
  const initialState = reducer(undefined, {});
  deepFreeze(initialState);
  deepFreeze(expectedState);
  let stateType = 'non-object';
  if (Array.isArray(initialState)) stateType = 'array';
  else if (typeof initialState === 'object') stateType = 'object';

  // if single action, still put in an array
  const actions = Array.isArray(action) ? action : [action];

  let reduceStarter;
  if (stateType === 'array') reduceStarter = [...initialState, ...init];
  else if (stateType === 'object') reduceStarter = { ...initialState, ...init };

  // run all the actions/reducers and combine de states. We've to account for the fact that the initialState can be a literal, an object or an object
  const finalState = Object.keys(actions).reduce(
    (state, key) => reducer(state, actions[key]),
    reduceStarter,
  );

  if (stateType === 'array') expect(finalState).toEqual([...initialState, ...expectedState]);
  else if (stateType === 'object') {
    expect(finalState).toEqual({ ...initialState, ...expectedState });
  } else expect(finalState).toEqual(expectedState);
}
