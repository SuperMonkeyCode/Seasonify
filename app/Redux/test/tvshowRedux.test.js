import { testReducer } from '../../Libs/Testing';
import reducer, {
  tvshowActions as actions,
  tvshowSelectors as selectors,
  INITIAL_STATE,
} from '../tvshowRedux';

describe('Reducer', () => {
  it('provides initial state', () => {
    testReducer(reducer, {}, INITIAL_STATE);
    expect(reducer(undefined, {})).toMatchSnapshot();
  });

  it('handles addTvshow action', () => {
    testReducer(
      reducer,
      [
        actions.addTvshow({ id: 'abc-123', name: 'Deadwood' }),
        actions.addTvshow({ id: 'xyz-789', name: 'Breaking Bad' }),
      ],
      {
        'abc-123': { id: 'abc-123', name: 'Deadwood', seasons: {} },
        'xyz-789': { id: 'xyz-789', name: 'Breaking Bad', seasons: {} },
      },
    );
  });
});

describe('Selectors', () => {
  it('getTvshows', () => {
    expect(
      selectors.getTvshow({ tvshows: { a: { name: 'A' }, b: { name: 'B' } } }, 'b'),
    ).toEqual({ name: 'B' });
  });
});
