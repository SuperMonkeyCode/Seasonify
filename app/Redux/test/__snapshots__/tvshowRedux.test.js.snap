// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducer provides initial state 1`] = `
Object {
  "37a543ee-1ef2-4a03-b2d3-91ede4418089": Object {
    "allocine": 213,
    "id": "37a543ee-1ef2-4a03-b2d3-91ede4418089",
    "localizedName": null,
    "name": "Deadwood",
    "poster": "http://fr.web.img5.acsta.net/medias/nmedia/18/82/49/12/19623049.jpg",
    "seasons": Object {
      "1": Object {
        "allocine": 565,
        "episodes": 12,
        "id": 1,
        "year": 2004,
      },
      "2": Object {
        "3": Object {
          "allocine": 2664,
          "episodes": 12,
          "id": 3,
          "year": 2006,
        },
        "allocine": 711,
        "episodes": 12,
        "id": 2,
        "year": 2005,
      },
    },
    "year": 2004,
  },
}
`;
